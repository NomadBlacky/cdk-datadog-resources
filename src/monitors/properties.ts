/**
 * Credentials for the Datadog API
 */
export interface DatadogCredentials {
  /** Datadog API key */
  readonly ApiKey: string;
  /** Datadog application key */
  readonly ApplicationKey: string;
  /** Datadog API URL (defaults to https://api.datadoghq.com) Use https://api.datadoghq.eu for EU accounts. */
  readonly ApiURL?: string;
}

export type MonitorType =
  | 'composite'
  | 'event alert'
  | 'log alert'
  | 'metric alert'
  | 'process alert'
  | 'query alert'
  | 'service check'
  | 'synthetics alert'
  | 'trace-analytics alert';

export interface MonitorOptions {
  /** Whether or not to include a sample of the logs */
  readonly EnableLogsSample?: boolean;
  /** Message to include with a re-notification when renotify_interval is set */
  readonly EscalationMessage?: string;
  /** Time in seconds to delay evaluation */
  readonly EvaluationDelay?: number;
  /** Whether or not to include triggering tags into notification title */
  readonly IncludeTags?: boolean;
  /** Whether or not changes to this monitor should be restricted to the creator or admins */
  readonly Locked?: boolean;
  /** Number of locations allowed to fail before triggering alert */
  readonly MinLocationFailed?: number;
  /** Time in seconds to allow a host to start reporting data before starting the evaluation of monitor results */
  readonly NewHostDelay?: number;
  /** Number of minutes data stopped reporting before notifying */
  readonly NoDataTimeframe?: number;
  /** Whether or not to notify tagged users when changes are made to the monitor */
  readonly NotifyAudit?: boolean;
  /** Whether or not to notify when data stops reporting */
  readonly NotifyNoData?: boolean;
  /** Number of minutes after the last notification before the monitor re-notifies on the current status */
  readonly RenotifyInterval?: number;
  /** Whether or not the monitor requires a full window of data before it is evaluated */
  readonly RequireFullWindow?: boolean;
  /** ID of the corresponding synthetics check */
  readonly SyntheticsCheckID?: number;
  /** The threshold definitions */
  readonly Thresholds?: MonitorThresholds;
  /** The threshold window definitions */
  readonly ThresholdWindows?: MonitorThresholdWindows;
  /** Number of hours of the monitor not reporting data before it automatically resolves */
  readonly TimeoutH?: number;
}

export interface MonitorThresholds {
  /** Threshold value for triggering an alert */
  readonly Critical?: number;
  /** Threshold value for recovering from an alert state */
  readonly CriticalRecovery?: number;
  /** Threshold value for recovering from an alert state */
  readonly OK?: number;
  /** Threshold value for triggering a warning */
  readonly Warning?: number;
  /** Threshold value for recovering from a warning state */
  readonly WarningRecovery?: number;
}

export interface MonitorThresholdWindows {
  /** How long a metric must be anomalous before triggering an alert */
  TriggerWindow?: string;
  /** How long an anomalous metric must be normal before recovering from an alert state */
  RecoveryWindow?: string;
}
