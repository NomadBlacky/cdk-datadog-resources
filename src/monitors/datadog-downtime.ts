import { CfnResource, Construct } from '@aws-cdk/core';
import * as camelcaseKeys from 'camelcase-keys';
import { DatadogCredentials } from '../common/properties';

export interface DatadogDowntimeProps {
  /** Credentials for the Datadog API */
  readonly datadogCredentials: DatadogCredentials;

  /** Required scope */
  readonly scope: string;
}

/**
 * Datadog Monitors 2.0.0
 */
export class DatadogDowntime {
  constructor(scope: Construct, id: string, props: DatadogDowntimeProps) {
    const cfnProperties = camelcaseKeys(props, {
      deep: true,
      pascalCase: true,
    });
    new CfnResource(scope, id, {
      type: 'Datadog::Monitors::Downtime',
      properties: { ...cfnProperties },
    });
  }
}