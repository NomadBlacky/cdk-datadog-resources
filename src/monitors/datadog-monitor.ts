import { CfnResource, Construct } from '@aws-cdk/core';
// eslint-disable-next-line @typescript-eslint/no-require-imports
import camelcaseKeys = require('camelcase-keys');
import { MonitorOptions, MonitorType } from './properties';
import { DatadogCredentials } from '../common/properties';

export interface DatadogMonitorProps {
  /** Credentials for the Datadog API */
  readonly datadogCredentials: DatadogCredentials;

  /** The monitor query */
  readonly query: string;

  /** The type of the monitor */
  readonly type: MonitorType;

  /** A message to include with notifications for the monitor*/
  readonly message?: string;

  /** Name of the monitor */
  readonly name?: string;

  /** Tags associated with the monitor */
  readonly tags?: string[];

  /** The monitor options*/
  readonly options?: MonitorOptions;

  /** Whether or not the monitor is multi alert */
  readonly multi?: boolean;
}

/**
 * Datadog Monitor 3.0.0
 */
export class DatadogMonitor {
  constructor(scope: Construct, id: string, props: DatadogMonitorProps) {
    const cfnProperties = camelcaseKeys(props, {
      deep: true,
      pascalCase: true,
    });
    new CfnResource(scope, id, {
      type: 'Datadog::Monitors::Monitor',
      properties: { ...cfnProperties },
    });
  }
}
