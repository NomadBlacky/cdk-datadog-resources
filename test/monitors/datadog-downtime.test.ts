import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { DatadogDowntime } from '../../src/monitors/datadog-downtime';

test('Snapshot test', () => {
  const stack = new Stack();

  new DatadogDowntime(stack, 'TestMonitor', {
    datadogCredentials: {
      apiKey: 'DATADOG_API_KEY',
      applicationKey: 'DATADOG_APP_KEY',
    },
    scope: ["host:myserver", "service:myservice"],
    start: 1624542715,
    end: 1624546321
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});