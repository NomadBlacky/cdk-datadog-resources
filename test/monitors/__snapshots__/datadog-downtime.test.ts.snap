// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
Object {
  "Resources": Object {
    "TestMonitor": Object {
      "Properties": Object {
        "DatadogCredentials": Object {
          "ApiKey": "DATADOG_API_KEY",
          "ApplicationKey": "DATADOG_APP_KEY",
        },
        "End": 1624546321,
        "Scope": Array [
          "host:myserver",
          "service:myservice",
        ],
        "Start": 1624542715,
      },
      "Type": "Datadog::Monitors::Downtime",
    },
  },
}
`;
